ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -216,6 +241,158 @@ KEXT_STATIC void UseMainForkIfNamedStream(
+    {
+        if (newMaxPendingRenames > s_maxPendingRenames)
+        {
+            Array_CopyElements(newArray, s_pendingRenames, s_maxPendingRenames);
+            Array_DefaultInit(newArray + s_maxPendingRenames, newMaxPendingRenames - s_maxPendingRenames);
+            
+            arrayToFree = s_pendingRenames;
+            arrayToFreeLength = s_maxPendingRenames;
+            
+            s_pendingRenames = newArray;
+            s_maxPendingRenames = newMaxPendingRenames;
+        }
+        else
+        {
+            arrayToFree = newArray;
+            arrayToFreeLength = newMaxPendingRenames;
+        }
+    }
+    SpinLock_Release(s_renameLock);
+    
+    if (arrayToFree != nullptr)
+    {
+        assert(arrayToFreeLength > 0);
+        Memory_FreeArray(arrayToFree, arrayToFreeLength);
+    }
+}
+
+static void RecordPendingRenameOperation(vnode_t vnode)
+{
+    assertf(version_major >= 18, "KAUTH_FILEOP_WILL_RENAME is only supported by Darwin 18 and newer (version_major = %u)", version_major);
+    thread_t myThread = current_thread();
+    
+    bool resizeTable;
+    do
+    {
+        resizeTable = false;
+        uint32_t resizeTableLength = 0;
+        
+        SpinLock_Acquire(s_renameLock);
+        {
+            if (s_pendingRenameCount < s_maxPendingRenames)
+            {
+                s_pendingRenames[s_pendingRenameCount].thread = myThread;
+                s_pendingRenames[s_pendingRenameCount].vnode = vnode;
+                ++s_pendingRenameCount;
+            }
+            else
+            {
+                for (uint32_t i = 0; i < s_pendingRenameCount; ++i)
+                {
+                    assert(s_pendingRenames[i].thread != myThread);
+                }
+
+                resizeTable = true;
+                
+                resizeTableLength = MAX(1u, MIN(UINT32_MAX, s_maxPendingRenames * UINT64_C(2)));
+                assert(resizeTableLength > s_maxPendingRenames);
+            }
+        }
+        SpinLock_Release(s_renameLock);
+        
+        if (resizeTable)
+        {
+            ResizePendingRenames(resizeTableLength);
+        }
+    } while (resizeTable);
+}
+
+static bool DeleteOpIsForRename(vnode_t vnode)
+{
+    if (version_major < 18)
+    {
+        // High Sierra and earlier do not support WILL_RENAME notification, so we have to assume any delete may be caused by a rename
+        assert(s_pendingRenameCount == 0);
+        assert(s_maxPendingRenames == 0);
+        return true;
+    }
+    
+    bool isRename = false;
+    
+    thread_t myThread = current_thread();
+    
+    SpinLock_Acquire(s_renameLock);
+    {
+        for (uint32_t i = 0; i < s_pendingRenameCount; ++i)
+        {
+            if (s_pendingRenames[i].thread == myThread)
+            {
+                isRename = true;
+                assert(s_pendingRenames[i].vnode == vnode);
+                --s_pendingRenameCount;
+                if (i != s_pendingRenameCount)
```

Comment:
The idea here is that you lower pendingRenameCount by one, and swap the new free index with what was at the end.

Code looks correct, but it's a lot to take in.  Some comments here would be helpful.

Reply:
I'd really love a unit test here.  One that explicity swaps a vnode not at the end.

Reply:
I've now added [`testPendingRenamesOutOfOrderInsertAndRemoval`](https://github.com/microsoft/VFSForGit/pull/1193/commits/753ff3236def50cc95bbfec98480b1f50f2c35d5#diff-3da2dbea50d618b6315c407d65c25658R677) which covers this situation.